(TK-391) Add unit test for overflow event handling

justinstoller opened the pull request:
Previously we tested overflow handling at the :integration level. This
has the unfortunate quality of requiring us to cause an error state
within Java/inotify. Up until this point we have been able to do this
easily without major repercussions on the systems that run the test.

That is not true in CI however, where we have a version of the JDK that
will not easily throw overflow events. In fact going to the lengths
required to cause an overflow event on these hosts is just as likely to
crash the JVM or fill up the disk.

This patch adds a unit test to prove our handling of overflow events is
satisfactory, and refactors the production code to make it unit
testable.

justinstoller committed:
(TK-391) Add unit test for overflow event handling

Previously we tested overflow handling at the :integration level. This
has the unfortunate quality of requiring us to cause an error state
within Java/inotify. Up until this point we have been able to do this
easily without major repercussions on the systems that run the test.

That is not true in CI however, where we have a version of the JDK that
will not easily throw overflow events. In fact going to the lengths
required to cause an overflow event on these hosts is just as likely to
crash the JVM or fill up the disk.

This patch adds a unit test to prove our handling of overflow events is
satisfactory, and refactors the production code to make it unit
testable.

justinstoller commented:
This should fix our current issues in CI

haus commented on test/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_service_test.clj:
I think making this `(fn [func] (func))` might make it a little more readable.

haus commented on test/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_service_test.clj:
Would be good to document that this is implementing an interface that is otherwise private to us (as Event is a private static class inside of an abstract class) https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/sun/nio/fs/AbstractWatchKey.java#L190-L222

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -87,33 +87,42 @@
                                           (filter dir-create?)
                                           (map #(.toPath (:path %)))))))
 
+(defn retrieve-events
```
Having a schema here would be cool.

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -87,33 +87,42 @@
+        events (.pollEvents watch-key)]
+    [watch-key events]))
+
+(defn process-events
```
Same as above, using a schema would make it clear what is going on.

haus replied:
Also, this might be better called `process-events!` as it is all about the side effects.

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -87,33 +87,42 @@
+                (trs "Events:")
+                (pprint-events clojure-events))
+    (log/tracef "%s\n%s"
+                (trs "orig-events:")
```
Would be good to settle on orig-events or java-events and just have one. Having both means trying to figure out which clojure var the "orig-events" string is referring to. I'd vote to stick with orig-events, as users seeing logs probably aren't going to care about the java part of the event as much as that it is a preprocessed event.

haus commented on test/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_service_test.clj:
Having at least one `testing` call here to describe what is going on would be good.

haus commented:
This looks pretty good. I left some schema/naming/documentation comments, but once those are resolved I'm good merging this.

justinstoller committed:
(TK-391) Clean up and documentation

justinstoller commented:
D'oh! I was so excited at getting the damned thing to work I opened the PR without going back and updating the fns with schemas and docs. I believe I've addressed all your feedback, @haus

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -87,33 +87,48 @@
                                           (filter dir-create?)
                                           (map #(.toPath (:path %)))))))
 
+(schema/defn retrieve-events
```
Would it make sense to add a schema here for return type?

justinstoller committed:
(TK-391) Add schema to retrieve-events

justinstoller committed:
(TK-391) remove unneeded with-test-logging from overflow test

haus commented:
Alright, I'm in. :+1:
